src/cmd/cmd_opt.ml

```diff
@@ -7,11 +7,15 @@ open Syntax
 let optimize_file ~unsafe filename =
   Compile.File.until_optimize ~unsafe ~optimize:true filename
 
-let cmd debug unsafe files =
+let print_or_emit ~unsafe file outfile =
+  let+ m = optimize_file ~unsafe file in
+  let m = Binary_to_text.modul m in
+  match outfile with
+  | Some name -> Bos.OS.File.writef name "%a@\n" Text.pp_modul m
+  | None -> Ok (Fmt.pr "%a@\n" Text.pp_modul m)
+
+let cmd debug unsafe file outfile =
   if debug then Log.debug_on := true;
-  list_iter
-    (fun file ->
-      let+ m = optimize_file ~unsafe file in
-      let m = Binary_to_text.modul m in
-      Fmt.pr "%a@\n" Text.pp_modul m )
-    files
+  match print_or_emit ~unsafe file outfile with
+  | Error _ -> raise Exit
+  | Ok _ -> Ok ()
```

Comment:
It should work if you simply remove the match, right?

Reply:
No it's not, I replace this with

```ocaml
Result.bind (print_or_emit ~unsafe file outfile) (fun _ -> Ok ())
```

let me know if it's okay.

Reply:
You have to replace:

```ocaml
let print_or_emit ~unsafe file outfile =
  let+ m = optimize_file ~unsafe file in
```

By:

```ocaml
let print_or_emit ~unsafe file outfile =
  let* m = optimize_file ~unsafe file in
```

Because either `writef` or your `Ok ...` are already returning a result. Basically, `let+` is like `let*` (binding in the monad) but also wraps the result in an `Ok` (map in the monad).

Once you did this, the return type of `print_or_emit` should be a proper Result (right now I believe it should be a Result wrapped in another Result...).

Then, you should be able to remove your `Result.bind` :)